Remove source on Import(IEnumerable<string> paths), Import(string path)

revam opened the pull request:
- Won't throw on unsupported beatmap
- Deletes file __only__ on a successful import.
- Writes to log if import or removal fails.

Alternative for removal of source file on import, also handles errors. Replaces #245.

revam committed:
Delete file after importing BeatmapSet w/error handling,

also changed batch-commit of multiple sets to database.

revam committed:
Update tests

revam committed:
Only delete on successful import.

peppy commented on osu.Game/Database/BeatmapDatabase.cs:
```diff
@@ -118,12 +118,43 @@ public void Reset()
+            Import(new [] { path });
+        }
+
+        private BeatmapSetInfo importBeatmapSet(string path)
```
weird name as this doesn't actually perform the database import (just copying of file and checking for pre-existing). should also document the null return on already-existing.

revam replied:
Would it be better to rename it to `getBeatmapSet` then? Also will document in next commit.

peppy commented on osu.Game/Database/BeatmapDatabase.cs:
```diff
@@ -188,7 +219,7 @@ public void Import(string path)
+            return beatmapSet;
         }
 
         public void Import(IEnumerable<BeatmapSetInfo> beatmapSets)
```
seems as though this should not be public, as the only way we can guarantee things are in a good state is if this is called by another method above inside of `BeatmapDatabase`. Might even work better to combine this with the `Import(paths)` method above.

revam replied:
I left it as-is because it was used in `TestCasePlaySongSelect`, but if desired I can combine it with `Import(paths)`.

revam committed:
Only set StoryBoardFile once

revam committed:
Document private method getBeatmapSet(path)

peppy commented on osu.Game/Database/BeatmapDatabase.cs:
```diff
@@ -118,12 +118,49 @@ public void Reset()
+
+        /// <summary>
+        /// Duplicates content from <paramref name="path"/> to storage and returns a representing <see cref="BeatmapSetInfo"/>.
+        /// Returns null if a representation already exists.
```
how about instead of returning `null`, we return the existing BeatmapSetInfo and then check in the actual Import method whether a databased ID is present (and don't insert again)? should simplify this logic.

revam committed:
Only add set if ID is 0

revam committed:
Update summary and commentary

revam committed:
Merged branch master into remove-source-on-import

revam committed:
Ensure correct exception logged

revam committed:
Added new test

When file is in use, we expect it to presist removal, in other words, fail to delete.

revam committed:
More summaries

peppy committed:
Add comment about .ID == 0 check.

peppy committed:
Merge branch 'master' into remove-source-on-import
